godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,818 @@
+/// can then be returned from [`IScriptExtension::instance_create`](crate::engine::IScriptExtension::instance_create).
+pub trait ScriptInstance {
+    /// Name of the new class the script implements.
+    fn class_name(&self) -> GString;
+
+    /// Property setter for Godots virtual dispatch system. The engine will call this function when it wants to change a property on the script.
+    fn set(&mut self, name: StringName, value: &Variant) -> bool;
+
+    /// Property getter for Godots virtual dispatch system. The engine will call this function when it wants to read a property on the script.
+    fn get(&self, name: StringName) -> Option<Variant>;
+
+    /// A list of all the properties a script exposes to the engine.
+    fn property_list(&self) -> Rc<Vec<PropertyInfo>>;
+
+    /// A list of all the methods a script exposes to the engine.
+    fn method_list(&self) -> Rc<Vec<MethodInfo>>;
+
+    /// Method invoker for Godots virtual dispatch system. The engine will call this function when it wants to call a method on the script.
+    ///
+    /// All method calls are taking a mutable reference of the script instance, as the engine does not differentiate between immutable and
+    /// mutable method calls like rust.
+    ///
+    /// It's important that the script does not cause a second call to this function while executing a method call. This would result in a panic.
+    fn call(
+        &mut self,
+        method: StringName,
+        args: &[&Variant],
+    ) -> Result<Variant, sys::GDExtensionCallErrorType>;
+
+    /// Identifies the script instance as a placeholder. Godot treats placeholder instances slightly different than normal instances.
+    fn is_placeholder(&self) -> bool;
+
+    /// Validation function for the engine to verify if the script exposes a certain method.
+    fn has_method(&self, method: StringName) -> bool;
+
+    /// Lets the engine get a reference to the script this instance was created for.
+    fn get_script(&self) -> Gd<Script>;
+
+    /// Lets the engine fetch the type of a particular property.
+    fn get_property_type(&self, name: StringName) -> sys::VariantType;
+
+    /// String representation of the script instance.
+    fn to_string(&self) -> GString;
+
+    /// Lets the engine get the owner / base object the script is attached to.
+    fn owner(&self) -> Gd<Object>;
+
+    /// A dump of all property names and values that are exposed to the engine.
+    fn property_state(&self) -> Vec<(StringName, Variant)>;
+
+    /// Lets the engine get a reference to the [`ScriptLanguage`] this instance belongs to.
+    fn language(&self) -> Gd<ScriptLanguage>;
+
+    /// Callback from the engine when the reference count of the base object has been decreased.
+    fn refcount_decremented(&self) -> bool;
+
+    /// Callback from the engine when the reference count of the base object has been increased.
+    fn refcount_incremented(&self);
```

Comment:
I would name those `on_refcount_{incremented|decremented}`.

The first method should have a documentation about the return value.

Reply:
renamed the functions and added documentation.